### PR TITLE
Params refactor, simplified

### LINKED ISSUE #1041: Refactor params to use the same underlying implementation -- $200 bounty
Currently there are two (or maybe three in Java):

* selfdrive/common/params.cc
* common/params.py

$200 for a merged PR that follows the spec below.

### PR DESCRIPTION
**Pull request by pd0wm**
Simplified version of:  https://github.com/commaai/openpilot/pull/2298. Will see if can get away by simplifying the python api to what the c++ version already supports.

TODO:
- [x] Check on completely clean device and make sure folders are create sucessfully
- [x] Handle interrupt in get_db_value_blocking

### CODE REVIEW

File: common/params_pyx_setup.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import os
+from distutils.core import Extension, setup
+from Cython.Build import cythonize
+from common.cython_hacks import BuildExtWithoutPlatformSuffix
+from common.basedir import BASEDIR
+
+sourcefiles = ['params_wrapper.pyx']
+extra_compile_args = ["-std=c++17"]
```
Comment (adeebshihadeh) [inline common/params_pyx_setup.py:8]: why is this 17?

Comment (pd0wm) [reply common/params_pyx_setup.py:8]: Was done in the original params PR. Will change it what we use everywhere else. We should switch to 17 at some point though, it has some nicer for loops over maps.

**Comment by adeebshihadeh** [comment]
Let's add a test for TODO #1

File: common/params_pxd.pxd [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  pass
+
+# Declare the class with cdef
+cdef extern from "selfdrive/common/params.h":
+  cdef cppclass Params:
+    Params(bool)
+    Params(string)
+    string get(string, bool) nogil
+    int delete_db_value(const char*);
+    int write_db_value(string, string);
```
Comment (adeebshihadeh) [inline common/params_pxd.pxd:17]:
```suggestion
    int delete_db_value(const char*)
    int write_db_value(string, string)
```

Comment (adeebshihadeh) [reply common/params_pxd.pxd:17]: why does delete take a const char and everything else uses a string?

File: common/params_pyx.pyx [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+from libcpp cimport bool
+from libcpp.string cimport string
+from params_pxd cimport Params as c_Params
+
+cdef enum TxType:
+  PERSISTENT = 1
+  CLEAR_ON_MANAGER_START = 2
+  CLEAR_ON_PANDA_DISCONNECT = 3
+
+keys = {
```
Comment (adeebshihadeh) [inline common/params_pyx.pyx:13]: why are we only checking this in python?

File: common/params_pyx.pyx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+  def __dealloc__(self):
+    del self.p
+
+  def _clear_keys_with_type(self, tx_type):
+    for key in keys:
+      if tx_type in keys[key]:
+        self.delete(key)
+
+  def manager_start(self):
```
Comment (adeebshihadeh) [inline common/params_pyx.pyx:91]: we're missing a clear all function now. maybe we should just make a clear_keys function that takes a list of tx types and the default is all?

File: common/SConscript [hunk 5]
```diff
@@ -4,3 +4,11 @@ Import('env', 'cython_dependencies')
 env.Command(['common_pyx.so', 'clock.cpp'],
             cython_dependencies + ['common_pyx_setup.py', 'clock.pyx'],
             "cd common && python3 common_pyx_setup.py build_ext --inplace")
+
+# Build cython params module
+env.Command(['params_pyx.so', 'params_wrapper.cpp'],
```
Comment (adeebshihadeh) [inline common/SConscript:9]:
```suggestion
env.Command(['params_pyx.so', 'params_pyx.cpp'],
```

File: common/params_pyx.pyx [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  b"Offroad_UpdateFailed": [TxType.CLEAR_ON_MANAGER_START],
+}
+
+class UnknownKeyName(Exception):
+  pass
+
+cdef class Params:
+  cdef c_Params* p
+
+  def __cinit__(self, d=None):
```
Comment (adeebshihadeh) [inline common/params_pyx.pyx:77]: might as well add an arg for persistent params to be consistent with the c++ api

File: common/params_pxd.pxd (hunk 2 above)
Comment (pd0wm) [reply common/params_pxd.pxd]: Fixed. Think we should move everything in c++ over to the new std::string versions at some point and retire the pointer based ones.

File: common/params_pyx.pyx (hunk 3 above)
Comment (pd0wm) [reply common/params_pyx.pyx:13]: Do we really care? I could move it to C++ but not sure if it's worth the complexity.

**Comment by pd0wm** [comment]
Folders need to be created recursively. If we move openpilot to c++17 we can use std::filesystem

**Pull request merged**